app/upload.go

```diff
@@ -0,0 +1,241 @@
+	if err := us.IsValid(); err != nil {
+		return nil, err
+	}
+
+	channel, err := a.GetChannel(us.ChannelId)
+	if err != nil {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.incorrect_channel_id.app_error",
+			map[string]interface{}{"channelId": us.ChannelId}, "", http.StatusBadRequest)
+	}
+	if channel.DeleteAt != 0 {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.cannot_upload_to_deleted_channel.app_error",
+			map[string]interface{}{"channelId": us.ChannelId}, "", http.StatusBadRequest)
+	}
+
+	us, storeErr := a.Srv().Store.UploadSession().Save(us)
+	if storeErr != nil {
+		return nil, model.NewAppError("CreateUploadSession", "app.upload.create.save.app_error", nil, storeErr.Error(), http.StatusInternalServerError)
+	}
+
+	return us, nil
+}
+
+func (a *App) GetUploadSession(uploadId string) (*model.UploadSession, *model.AppError) {
+	us, err := a.Srv().Store.UploadSession().Get(uploadId)
+	if err != nil {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(err, &nfErr):
+			return nil, model.NewAppError("GetUpload", "app.upload.get.app_error",
+				nil, nfErr.Error(), http.StatusNotFound)
+		default:
+			return nil, model.NewAppError("GetUpload", "app.upload.get.app_error",
+				nil, err.Error(), http.StatusInternalServerError)
+		}
+	}
+	return us, nil
+}
+
+func (a *App) GetUploadSessionsForUser(userId string) ([]*model.UploadSession, *model.AppError) {
+	uss, err := a.Srv().Store.UploadSession().GetForUser(userId)
+	if err != nil {
+		return nil, model.NewAppError("GetUploadsForUser", "app.upload.get_for_user.app_error",
+			nil, err.Error(), http.StatusInternalServerError)
+	}
+	return uss, nil
+}
+
+func (a *App) UploadData(us *model.UploadSession, rd io.Reader) (*model.FileInfo, *model.AppError) {
+	// prevent more than one caller to upload data at the same time for a given upload session.
+	// This is to avoid possible inconsistencies.
+	a.Srv().uploadLockMapMut.Lock()
+	locked := a.Srv().uploadLockMap[us.Id]
+	if locked {
+		// session lock is already taken, return error.
+		a.Srv().uploadLockMapMut.Unlock()
+		return nil, model.NewAppError("UploadData", "app.upload.upload_data.concurrent.app_error",
+			nil, "", http.StatusBadRequest)
+	}
+	// grab the session lock.
+	a.Srv().uploadLockMap[us.Id] = true
+	a.Srv().uploadLockMapMut.Unlock()
+
+	// reset the session lock on exit.
+	defer func() {
+		a.Srv().uploadLockMapMut.Lock()
+		delete(a.Srv().uploadLockMap, us.Id)
+		a.Srv().uploadLockMapMut.Unlock()
+	}()
+
+	// make sure it's not possible to upload more data than what is expected.
+	lr := &io.LimitedReader{
+		R: rd,
+		N: us.FileSize - us.FileOffset,
+	}
+	var err *model.AppError
+	var written int64
+	if us.FileOffset == 0 {
+		// new upload
+		written, err = a.WriteFile(lr, us.Path)
+		if written < minFirstPartSize && written != us.FileSize {
```

Comment:
Not security-related, but this can cause cryptic error messages when `us.FileSize` is smaller than `minFirstPartSize`.

```suggestion
		if written < minFirstPartSize && written != us.FileSize && us.FileSize < minFirstPartSize {
```

Reply:
The thing here is that we cannot resume (and should return error) uploads whose first segment/part is less than `minFirstPartSize` but we still want to support uploading files smaller than `minFirstPartSize` if they are done all at once (no resume), hence the need for an equality comparison.

The limited reader should already guarantee that `written` is never greater than `us.FileSize` and otherwise fail.

Reply:
I think you misread the suggested change

Reply:
Trying to clarify a bit; The current code in this PR returns the "first part too small" error in the case where `WriteFile` fails and writes nothing, including when the entire file is smaller than the minimum part size. It should return the real error instead, because the "too small" error doesn't make sense in that case.

By checking for `us.FileSize > minFirstPartSize` instead of `written != us.FileSize` we're excluding the small file case from this code path, but `written < minFirstPartSize && us.FileSize > minFirstPartSize` also implies `written != us.FileSize`.

Reply:
Handling the error first sounds good to me.

Reply:
It should be improved now. 
Let me know if that works and thanks for the precious feedback on this :+1:

Reply:
Yeah looks good to me